test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/GraphNodeTests.cs

```diff
@@ -0,0 +1,97 @@
+namespace NuGet.DependencyResolver.Core.Tests
+{
+    public class GraphNodeTests
+    {
+        [Theory]
+        [InlineData(null, true)]
+        [InlineData(false, false)]
+        [InlineData(true, true)]
+        public void GraphNode_WithDifferentHasEmptyParentNodes_CreateParentNodesCorrectly(bool? hasParentNodes, bool parentNodesHasNode)
+        {
+            LibraryRange libraryRangeA = GetLibraryRange("A", "1.0.0", "1.0.0");
+            LibraryRange libraryRangeB = GetLibraryRange("B", "1.0.0", "1.0.0");
+
+            GraphNode<RemoteResolveResult> nodeA, nodeB;
+
+            if (!hasParentNodes.HasValue)
+            {
+                nodeA = new GraphNode<RemoteResolveResult>(libraryRangeA);
+                nodeB = new GraphNode<RemoteResolveResult>(libraryRangeB);
+            }
+            else
+            {
+                nodeA = new GraphNode<RemoteResolveResult>(libraryRangeA, hasInnerNodes:true, hasParentNodes.Value);
+                nodeB = new GraphNode<RemoteResolveResult>(libraryRangeB, hasInnerNodes:true, hasParentNodes.Value);
+            }
+
+            bool sameParentNodes = (nodeA.ParentNodes == nodeB.ParentNodes);
+            if (!parentNodesHasNode)
+            {
+                //The ParentNodes should be pointing to the static EmptyList.
+                sameParentNodes.Should().BeTrue(because: "nodeA.ParentNodes and nodeB.ParentNodes should both point to the static EmptyList");
+
+                //EmptyList is immutable.
+                var exception = Assert.ThrowsAny<NotSupportedException>(
+                () => nodeA.ParentNodes.Add(nodeB));
+                Assert.Contains("of a fixed size.", exception.Message);
+            }
+            else
+            {
+                sameParentNodes.Should().BeFalse(because: "nodeA.ParentNodes and nodeB.ParentNodes should not point to the static EmptyList");
+            }
+        }
+
+        [Theory]
+        [InlineData(null, true)]
+        [InlineData(false, false)]
+        [InlineData(true, true)]
+        public void GraphNode_WithDifferentHasEmptyInnerNodes_CreateInnerNodesCorrectly(bool? hasInnerNodes, bool innerNodesHasNode)
+        {
+            LibraryRange libraryRangeA = GetLibraryRange("A", "1.0.0", "1.0.0");
+            LibraryRange libraryRangeB = GetLibraryRange("B", "1.0.0", "1.0.0");
+
+            GraphNode<RemoteResolveResult> nodeA, nodeB;
+
+            if (!hasInnerNodes.HasValue)
+            {
+                nodeA = new GraphNode<RemoteResolveResult>(libraryRangeA);
+                nodeB = new GraphNode<RemoteResolveResult>(libraryRangeB);
+            }
+            else
+            {
+                nodeA = new GraphNode<RemoteResolveResult>(libraryRangeA, hasInnerNodes.Value, hasParentNodes : true);
+                nodeB = new GraphNode<RemoteResolveResult>(libraryRangeB, hasInnerNodes.Value, hasParentNodes : true);
+            }
+
+            bool sameInnerNodes = (nodeA.InnerNodes == nodeB.InnerNodes);
```

Comment:
Same as above.

Reply:
Make sense! Fixed by comparing to Array.Empty directly.